l1-contracts/contracts/upgrades/BytecodesSupplier.sol

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.24;
+
+import {L2ContractHelper} from "../common/libraries/L2ContractHelper.sol";
+import {BytecodeAlreadyPublished} from "../common/L1ContractErrors.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @notice Contract that is used to track published L2 bytecodes.
+/// It will be the contract to which the preimages for the factory dependencies protocol upgrade transaction
+/// will be submitted to.
+/// @dev The contract has no access control as anyone is allowed to publish any bytecode.
+contract BytecodesSupplier {
+    /// @notice Event emitted when a bytecode is published.
+    event BytecodePublished(bytes32 indexed bytecodeHash, bytes bytecode);
+
+    /// @notice Mapping of bytecode hashes to the block number when they were published.
+    mapping(bytes32 bytecodeHash => uint256 blockNumber) public publishingBlock;
+
+    /// @notice Publishes the bytecode hash and the bytecode itself.
+    /// @param _bytecode Bytecode to be published.
+    function publishBytecode(bytes calldata _bytecode) public {
```

Comment:
Just to make sure EthWatch picks this up?

Reply:
The server does work with it (tested locally and briefly reviewed by Misha). The scheme works roughly like that:

- This contract is the place where we publish bytecodes. Just publishing does not make them known on L2.
- When we receive an upgrade, we see that there is a transaction with it, but it contains only hashes for factory deps to be marked as known prior to the execution of the upgrade. It also must contain default aa hash/bootloader hash in the array.
- So basically the bytecodes are marked as known only once the upgrade tx is executed. The contract in question mainly serve just as a place for the server to search for the actual preimages